### PR TITLE
[PLT-6712/PLT-7182] /remove and /kick slash command: Remove a member from channel

### PR DESCRIPTION
**Pull request by cpanato**
#### Summary
add slash command /kick and /remove to remove users from channel
WEBAPP PR: https://github.com/mattermost/mattermost-webapp/pull/159

previous PR: https://github.com/mattermost/mattermost-server/pull/7124

#### Ticket Link
JIRA: https://mattermost.atlassian.net/browse/PLT-6712 and https://mattermost.atlassian.net/browse/PLT-7182
GitHub: mattermost#6557 and mattermost#6975

#### Checklist
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json) updates

### CODE REVIEW

**Comment by lindalumitchell** [comment]
@jasonblais @cpanato I ran through the testing on this, and almost everything passed! Just a couple questions / issues:

1. I'm seeing double system messages after /remove or /kick is successful (the `Only visible to you` is posting separately, in different order depending on if @ sign was used):
- With @ sign before username: `user was removed from channel` first, then separate `(Only visible to you)` message
- Without @ sign before username: `(Only visible to you)` first, then separate `user was removed from the channel`
double-system-message

2. The number on the member icon increments down immediately, but username still appears in drop-down member list until after refresh (public and private channels, /remove and /kick). If I use Manage Members to remove a member, it is removed from the list immediately
removed-still-in-member-list

3. I'm still seeing that a non-admin is able to successfully /remove and /kick users from a private channel when `Policy > Enable managing of private channel members for:` set to `Team and System Admins` or just `System Admins`. The expected behavior is to display an error ephemeral message `You do not have the appropriate permissions to remove the member.` and the user is not removed from the channel

Further testing notes for reference:
Passed, both /remove and /kick:
- DM or GM: "You can't remove someone from a direct message channel."
- Not sure if expected (seems useful actually and I don't know why not), but it works on own username (redirected to Town Square)
- User not in channel: "delete1 is not a member of this channel."
- Not a user on the channel: "We couldn't find the user"
- Additional text after the username (after a space) is ignored; removal is successful
- Attachment on the post is ignored; removal is successful
- Town Square: "Cannot remove user from the default channel town-square"

**Comment by jasonblais** [comment]
@cpanato Can you help take a look at 1 - 3 above?

@lindalumitchell great work testing this

**Comment by cpanato** [comment]
sure, will dig into tomorrow if is that ok @jasonblais 
thanks @lindalumitchell for the testing :D

**Comment by jasonblais** [comment]
@cpanato Of course! Not urgent :)

**Comment by lindalumitchell** [comment]
@cpanato it looks like my first issue above is actually on master, not from this PR. @jasonblais filed a ticket for it: https://mattermost.atlassian.net/browse/PLT-7787

**Comment by cpanato** [comment]
yes, @lindalumitchell I told to him, and I have a fix for that, pushing in a minute :)
thanks for your catch!

**Comment by cpanato** [comment]
@lindalumitchell @jasonblais 
- item 1 is fixed in this PR https://github.com/mattermost/mattermost-server/pull/7572
- item 2 I still cannot find why sometimes it is not update the list, sometimes update and sometimes not. the code trigger a websocket which is the same used by the manage members.
- item 3 working on that.

**Comment by esethna** [comment]
@cpanato let us know if we can help at all with the remaining items above.

**Comment by cpanato** [comment]
@esethna I will need some support for the item 2 because i did not find the issue

**Comment by jasonblais** [comment]
@cpanato Maybe the code that triggers a websocket connection used by the manage members modal has a bug?

**Comment by jasonblais** [comment]
@cpanato If you're not sure, you may consider posting in the Developers channel on pre-release and someone might be able to help.

**Comment by cpanato** [comment]
@jasonblais # 1 and # 3 fixed, still looking into # 2, also posted in dev channel.

I will push on Monday because I forgot my yubico key at work and I cant commit without that lol.

**Comment by jwilander** [comment]
Right here https://github.com/mattermost/mattermost-webapp/blob/master/actions/websocket_actions.jsx#L382, you need to add:

```
dispatch({
    type: ChannelTypes.RECEIVED_PROFILE_NOT_IN_CHANNEL,
    data: {user_id: msg.broadcast.user_id},
    id: channelId
});
```

Essentially the redux store was not getting the information that user was removed from the channel. Dispatching this event should cover that

**Comment by cpanato** [comment]
@jwilander Thanks for the insight! one quick question, this `RECEIVED_PROFILE_NOT_IN_CHANNEL` does not exist in the redux repo, should I add that there as well or can we use `LEAVE_CHANNEL`? thanks!

**Comment by jwilander** [comment]
Sorry typo in my provided code, it exists under UserTypes https://github.com/mattermost/mattermost-redux/blob/master/src/action_types/users.js#L137

**Comment by cpanato** [comment]
@jasonblais @lindalumitchell @jwilander made the changes and add a PR for webapp: https://github.com/mattermost/mattermost-webapp/pull/159

thanks for all your support

**Comment by cpanato** [comment]
@jasonblais @jwilander do I need to link the mm-webapp pr to this pr in the code?

**Comment by jasonblais** [comment]
Confimed 1 and 3 fixed. 2 will be tested with https://github.com/mattermost/mattermost-webapp/pull/159 once this PR has merged.

Moving to dev review next.

@cpanato No need to link the webapp PR in the code.

File: app/command_remove.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	return doCommand(a, args, message)
+}
+
+func doCommand(a *App, args *model.CommandArgs, message string) *model.CommandResponse {
+	channel, err := a.GetChannel(args.ChannelId)
+	if err != nil {
+		return &model.CommandResponse{Text: args.T("api.command_channel_rename.channel.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+	}
+
+	if channel.Type == model.CHANNEL_OPEN && !a.SessionHasPermissionToChannel(args.Session, args.ChannelId, model.PERMISSION_MANAGE_PUBLIC_CHANNEL_PROPERTIES) {
```
Comment (ccbrown) [inline app/command_remove.go]: Should this be `model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS`?

Comment (cpanato) [reply app/command_remove.go]: thats true, thanks!! done @ccbrown

**Pull request merged**